src/nfa/intersection.cc

```diff
@@ -16,177 +16,254 @@
-            {
-                for (const State other_state_to: moves[1]->targets)
-                {
-                    create_product_state_and_trans(
-                            product, product_map, lhs, rhs, pairs_to_process,
-                            this_state_to, other_state_to, intersection_transition
-                    );
+            Symbol symbol = same_symbol_posts[0]->symbol;
+            if (symbol < first_epsilon) {
+                SymbolPost product_symbol_post{symbol};
+                for (const State lhs_target: same_symbol_posts[0]->targets) {
+                    for (const State rhs_target: same_symbol_posts[1]->targets) {
+                        create_product_state_and_move(lhs_target, rhs_target, product_symbol_post);
+                    }
                 }
+                //add_product_symbol_post(product, pair_to_state, lhs_source, rhs_source, product_symbol_post);
+                StatePost &product_state_post{product.delta.mutable_state_post(get_product_state(lhs_source,rhs_source))};
+                product_state_post.push_back(std::move(product_symbol_post));
             }
-            add_product_transition(product, product_map, pair_to_process, intersection_transition);
+            else
+                break;
         }
 
-        if (preserve_epsilon) {
-            // Add transitions of the current state pair for an epsilon preserving product.
-
-            // Check for lhs epsilon transitions.
-            const StatePost& lhs_post{ lhs.delta[pair_to_process.first] };
-            if (!lhs_post.empty()) {
-                const auto& lhs_state_last_transitions{ lhs_post.back() };
-                if (epsilons.find(lhs_state_last_transitions.symbol) != epsilons.end()) {
-                    // Compute product for state transitions with lhs state epsilon transition.
-                    // Create transition from the pair_to_process to all pairs between states to which first transition
-                    //  goes and states to which second one goes.
-                    SymbolPost intersection_transition{ lhs_state_last_transitions.symbol };
-                    for (const State lhs_state_to: lhs_state_last_transitions.targets) {
-                        create_product_state_and_trans(product, product_map, lhs, rhs, pairs_to_process,
-                                                       lhs_state_to, pair_to_process.second,
-                                                       intersection_transition);
-                    }
-                    add_product_transition(product, product_map, pair_to_process, intersection_transition);
+        // Add epsilon transitions, from lhs e-transitions.
+        const StatePost& lhs_state_post{lhs.delta[lhs_source] };
+
+        //TODO: handling of epsilons might not be ideal, don't know, it would need some brain cycles to improve.
+        // (handling of normal symbols is more important and it is ok)
```

Comment:
Well I guess this is not true. In `noodler` we use heavily this epsilon product.

Reply:
Is there a benchmark with it at tacas?
Anyway, I currently don't know how to do it better. 
Lets think about it if we find it slow on some benchmark.

Reply:
Well, depends on which tacas paper do you mean. For `mata` we don't use it.